### PR TITLE
Feat: bearer token authentication support

### PR DESCRIPTION
**Pull request by skvark**
### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you installed `pre-commit` with `pip3 install pre-commit` and set up hooks with `pre-commit install`?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

## What

This pull request adds a custom Bearer authentication token provider function for the HTTP clients. Additionally, Bearer token authentication is enabled for the GRPC clients as well.

## Why

Currently, it is impossible to use the clients properly with token-based authentication, since the the only way of passing the token is via the static `metadata` parameter. When the token expires, the whole client needs to be recreated.

## Changes

The clients have a new parameter, `auth_token_provider`. This can be sync or async function, that returns `str` Bearer token.

For the HTTP clients, this PR adds a new custom `BearerAuth` class that is based on `httpx.Auth` class (https://www.python-httpx.org/advanced/authentication/#custom-authentication-schemes). This is passed down in the `auth` parameter to the `httpx` library.

The GRPC code in `connection.py` has been refactored to support the new feature and the code in `get_channel` and `get_async_channel`  should be simpler now.

## Example

There are simple examples in the tests, but here is a proper example with `azure-identity` library:

```python
import qdrant_client
from azure.identity.aio import DefaultAzureCredential

credential = DefaultAzureCredential(exclude_interactive_browser_credential=False)

async def get_token():
    result = await credential.get_token("your_qdrant_app_scope")
    return result.token

client = qdrant_client.AsyncQdrantClient(
    url=f"your.qdrant.app.address",
    https=True,
    port=443,
    auth_token_provider=get_token,
)

await client.get_collections()
```

### CODE REVIEW

**Review by joein — COMMENTED** [review]
@skvark thank you for the contribution!

**Comment by joein** [comment]
I've left some comments, I'll address them and merge

File: qdrant_client/async_qdrant_remote.py [hunk 1]
```diff
@@ -129,6 +137,16 @@ def __init__(
             self._rest_args["limits"] = limits
         if self._timeout is not None:
             self._rest_args["timeout"] = self._timeout
+        if self._auth_token_provider is not None:
+            if self.__class__.__name__ == "QdrantRemote" and asyncio.iscoroutinefunction(
```
[the remark this answers is not in the document]
Comment (skvark) [reply qdrant_client/async_qdrant_remote.py]: Yeah, I didn't like this myself either. It's better now after your changes.

File: qdrant_client/auth/bearer_auth.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import asyncio
+import threading
```
[the remark this answers is not in the document]
Comment (skvark) [reply qdrant_client/auth/bearer_auth.py]: Forgot to clean these after removing some stuff from the auth class.

File: qdrant_client/connection.py [hunk 3]
```diff
@@ -125,14 +126,18 @@ class _ClientAsyncCallDetails(
     pass
 
 
-def header_adder_interceptor(new_metadata: List[Tuple[str, str]]) -> _GenericClientInterceptor:
+def header_adder_interceptor(
+    new_metadata: List[Tuple[str, str]] = [],
```
[the remark this answers is not in the document]
Comment (skvark) [reply qdrant_client/connection.py]: True 👍

**Comment by skvark** [comment]
> I've left some comments, I'll address them and merge

Thanks for fixing the issues!

**Comment by joein** [comment]
Hey @skvark 
I've given this PR another thought

I am somewhat reluctant on bringing new parameters to init (and I think that auth_token_provider actually affects only a small subset of users)

Afaik it's not a common practice to propagate token reissue logic into a database object, and it is usually handled on the client side, isn't it?

**Comment by skvark** [comment]
> Hey @skvark I've given this PR another thought
> 
> I am somewhat reluctant on bringing new parameters to init (and I think that auth_token_provider actually affects only a small subset of users)
> 
> Afaik it's not a common practice to propagate token reissue logic into a database object, and it is usually handled on the client side, isn't it?

Instead of init, we could introduce some new method to the client that registers the authentication token provider.

Here's one example from SQLAlchemy. It uses an event-based approach, but the renewal mechanism is the same: https://docs.sqlalchemy.org/en/20/core/engines.html#generating-dynamic-authentication-tokens

Weaviate implements the OIDC auth flows themselves (but their DB also validates the tokens so it's a bit different): [Weaviate](https://github.com/weaviate/weaviate-python-client/blob/main/weaviate/connect/authentication.py)

All databases should sit behind a proper authentication (and authorization) layer in addition to firewalls. Oauth 2.0 bearer tokens are the standard way of implementing that layer. We can use the current QDrant client, but not having a proper way to update the authorization headers on the existing Qdrant database client object leads to unnecessarily complicated code on our side.

**Comment by joein** [comment]
I'm going to merge it now, thank you for the contribution @skvark

**Comment by skvark** [comment]
> I'm going to merge it now, thank you for the contribution @skvark

Thanks!

**Pull request merged**